### PR TITLE
Thingifapi builder

### PR DESCRIPTION
**Pull request by yongpingchen**
- Improve ThingIFAPI.Builder.build to check actionTypes and stateTypes
- Add tests for ThingIFAPI.Builder

### CODE REVIEW

File: thingif/src/main/java/com/kii/thingif/KiiApp.java [hunk 1]
```diff
@@ -50,7 +50,7 @@ public KiiApp(String appID, String appKey, Site site) {
      * @param hostName Host Name of the app.
      * @throws  IllegalArgumentException when appID, appKey and hostName is null or empty string.
      */
-    KiiApp(String appID, String appKey, String hostName) {
+    public KiiApp(String appID, String appKey, String hostName) {
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/KiiApp.java]: Can we construct `KiiApp` instance without builder?

File: thingif/src/test/java/com/kii/thingif/thingifapi/BuilderTest.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+                null, 
+                new KiiApp("appid", "appkey", Site.JP), 
+                new Owner(new TypedID(TypedID.Types.USER, "user1234"), "token"));
+    }
+    @Test(expected = IllegalArgumentException.class)
+    public void newBuilderWithNullAppIDTest() throws Exception {
+        ThingIFAPI.Builder.newBuilder(
+                this.context,
+                new KiiApp(null, "appkey", Site.JP),
+                new Owner(new TypedID(TypedID.Types.USER, "user1234"), "token"));
```
Comment (shinsuke-nara) [inline thingif/src/test/java/com/kii/thingif/thingifapi/BuilderTest.java]: This is Test for `KiiApp` constructor. `ThingIFAPI.Builder` is needless.

What do you think?

File: thingif/src/test/java/com/kii/thingif/thingifapi/BuilderTest.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    }
+    @Test(expected = IllegalArgumentException.class)
+    public void newBuilderWithEmptyAppIDTest() throws Exception {
+        ThingIFAPI.Builder.newBuilder(
+                this.context,
+                new KiiApp("", "appkey", Site.JP),
+                new Owner(new TypedID(TypedID.Types.USER, "user1234"), "token"));
```
Comment (shinsuke-nara) [inline thingif/src/test/java/com/kii/thingif/thingifapi/BuilderTest.java]: This is Test for `KiiApp` constructor. `ThingIFAPI.Builder` is needless.

What do you think?

File: thingif/src/test/java/com/kii/thingif/thingifapi/BuilderTest.java [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    }
+    @Test(expected = IllegalArgumentException.class)
+    public void newBuilderWithNullAppKeyTest() throws Exception {
+        ThingIFAPI.Builder.newBuilder(
+                this.context,
+                new KiiApp("appid", null, Site.JP),
+                new Owner(new TypedID(TypedID.Types.USER, "user1234"), "token"));
```
Comment (shinsuke-nara) [inline thingif/src/test/java/com/kii/thingif/thingifapi/BuilderTest.java]: This is Test for `KiiApp` constructor. `ThingIFAPI.Builder` is needless.

What do you think?

File: thingif/src/test/java/com/kii/thingif/thingifapi/BuilderTest.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    }
+    @Test(expected = IllegalArgumentException.class)
+    public void newBuilderWithEmptyAppKeyTest() throws Exception {
+        ThingIFAPI.Builder.newBuilder(
+                this.context,
+                new KiiApp("appid", "", Site.JP),
+                new Owner(new TypedID(TypedID.Types.USER, "user1234"), "token"));
```
Comment (shinsuke-nara) [inline thingif/src/test/java/com/kii/thingif/thingifapi/BuilderTest.java]: This is Test for `KiiApp` constructor. `ThingIFAPI.Builder` is needless.

What do you think?

File: thingif/src/test/java/com/kii/thingif/thingifapi/BuilderTest.java [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    }
+    @Test(expected = IllegalArgumentException.class)
+    public void newBuilderWithNullSiteTest() throws Exception {
+        ThingIFAPI.Builder.newBuilder(
+                this.context,
+                new KiiApp("appid", "appkey", (Site) null),
+                new Owner(new TypedID(TypedID.Types.USER, "user1234"), "token"));
```
Comment (shinsuke-nara) [inline thingif/src/test/java/com/kii/thingif/thingifapi/BuilderTest.java]: This is Test for `KiiApp` constructor. `ThingIFAPI.Builder` is needless.

What do you think?

File: thingif/src/test/java/com/kii/thingif/thingifapi/BuilderTest.java [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    public void newBuilderWithNullOwnerTest() throws Exception {
+        ThingIFAPI.Builder.newBuilder(
+                this.context,
+                new KiiApp("appid", "appkey", Site.JP), null);
+    }
+    @Test(expected = IllegalArgumentException.class)
+    public void newBuilderWithNullBaseUrlTest() throws Exception {
+        ThingIFAPI.Builder.newBuilder(
+                this.context,
+                new KiiApp("appid", "appkey", (String)null),
```
Comment (shinsuke-nara) [inline thingif/src/test/java/com/kii/thingif/thingifapi/BuilderTest.java]: This is Test for `KiiApp` constructor. `ThingIFAPI.Builder` is needless.

What do you think?

File: thingif/src/test/java/com/kii/thingif/thingifapi/BuilderTest.java [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+                new Owner(new TypedID(TypedID.Types.USER, "user1234"), "token"));
+    }
+    @Test(expected = IllegalArgumentException.class)
+    public void newBuilderWithEmptyBaseUrlTest() throws Exception {
+        ThingIFAPI.Builder.newBuilder(
+                this.context,
+                new KiiApp("appid", "appkey", ""),
```
Comment (shinsuke-nara) [inline thingif/src/test/java/com/kii/thingif/thingifapi/BuilderTest.java]: This is Test for `KiiApp` constructor. `ThingIFAPI.Builder` is needless.

What do you think?

File: thingif/src/main/java/com/kii/thingif/KiiApp.java (hunk 1 above)
Comment (yongpingchen) [inline thingif/src/main/java/com/kii/thingif/KiiApp.java]: @shinsuke-nara The Builder is only for build KiiApp from hostname, which can be very complex(As document of KiiApp.Builder describes). So I think the Builder is necessary.

File: thingif/src/test/java/com/kii/thingif/thingifapi/BuilderTest.java (hunk 2 above)
Comment (yongpingchen) [inline thingif/src/test/java/com/kii/thingif/thingifapi/BuilderTest.java]: May be move them to other KiiAppTest.java is better?

**Comment by yongpingchen** [comment]
> If ActionTypes and StateTypes are mandatory for ThingIFAPI. ThingIFAPI.Builder should receive them at newBuilder() method.
There are 2 `newBuilder` factory methods:
- one receives these 2 parameters. 

- one without these 2 parameters. For the purpose of use chain to register one by one

```java
ThingIFAPI.Builder builder = ThingIFAPI.Builder.
       newBuilder(context, app, ower)
       .registerActionType("airConditionerAlias", AirConditionerAction.class)
       .registerActionType(...);
```

**Pull request merged**